### ISSUE TITLE
Editor crashes when importing a large number of sprites

### ISSUE THREAD

**Issue by morganwillcock** [adventuregamestudio/ags#702]
Both the existing build of the Editor (tested on 3.4.3.1) and the work in progress one, will crash when too many sprites are imported and the project data is not saved. This will be an issue for bulk sprite re-import, to recreate a missing spritecache.

As a test, I imported the same image (720x450 32bit PNG) 500 times, and memory usage went from around 30 MB to around 700 MB. When saving the project, usage returns to 30MB. A second import of the same 500 images (for a total of 1000) will sometimes crash or sometimes complete. Typically the memory usage at the time of crashing is between 1.2 GB and 1.5 GB.

I've tried setting `<gcAllowVeryLargeObjects enabled="true"/>` in app.config, but this didn't make a noticeable difference to the memory usage at time of crashing.

A memory snapshot shows that the usage is down to AGS Native (you have to enabled mixed-mode in the memory profiler to see it at all).

![mem2](https://user-images.githubusercontent.com/14958747/55591048-7a16c180-572c-11e9-97a1-e9e5ca0c98b7.png)

**Comment by ivan-mogilko** [comment]
As a quick thought, guess the only proper way to handle this is to introduce periodical saving and freeing the cache or part of the cache. But assuming default sprite file is a part of the project and by design should not be saved unless user hits Save, then we need another temporary sprite file that is being appended and then copied into (or replaces) default file upon saving the project.

**Comment by morganwillcock** [comment]
I think it might be better to postpone the task of trying to recreate the spritecache from scratch and look at moving its build process into the game compiler (so project/design-time data is always separate from game data). So far (for version 3.5 of the editor) I don't think we have removed any functionality, only fixed or improved what was there. For this (and fixing edge cases for clipboard imports and room palettes, and concurrent editing of room data) I think it probably aligns better with a change in approach that gives a more straightforward solution for all of these issues.

I keep meaning to re-write my notes that were on the YouTrack server here, but would I just open that in a new issue or should I add it as a project?

**Comment by ivan-mogilko** [comment]
If I knew this is going to be that complicated I'd work on this too instead of wasting time on not so important things! This feature was one of the reasons I kept working on AGS in this year.

**Comment by ivan-mogilko** [comment]
Alright, so, we call it a failure and wrap the version up. We need to restructure the milestone list to know what else left to be fixed.

**Comment by ivan-mogilko** [comment]
I am not certain, but I think I am going to look into this when I'm done with my current tasks.
I know I sound crazy about this issue, but this was an important feature in my opinion, that would lift a part of burden from peoples shoulders who must keep sprite file under source control.
Without this working the most of 3.5.0 changes look like cosmetic improvement in comparison.

Moving to open data is of course right thing, but there's a lot to do and code to rewrite.
Meanwhile recreating spritefile is like couple of steps away. Maybe I will try to work on this as a small update to 3.5.0.

**Comment by morganwillcock** [comment]
I think to add it in the current state it would need:
* a fix for this specific issue, i.e. crashing when importing lots of sprites
* checksum of spritecache file that is written into project data so that the editor knows not to attempt loading sprites, or maybe if AGSNative has support for it there could be a checksum per sprite
* add editor preference for spritesheet validation, or assume validation based on the opposite of the current option (from existing source control integration, that flags the binary data to be added to source control)
* project load process should be possible without reading any sprite data (otherwise you have to switch in a placeholder just to get the point where the editor is functional, that is what [my initial test](https://github.com/morganwillcock/ags/tree/rebuild-spritecache) was doing - effectively just a sprite replace on every sprite, with a placeholder for every failed sprite load)

The majority of work to get to this point was adding sprite reload for spritesheet imports, and the Spritesheet class is just a rectangle generator that wraps import calls to the native proxy, so there has been no wasted effort in the direction of working better with source control. I appreciate that is it a useful, but it is also tagged as AGS4 on the issue tracker (https://github.com/adventuregamestudio/ags/issues/468) with data format change as a sub-task, and there is no [milestone](https://github.com/adventuregamestudio/ags/issues?q=is%3Aopen+is%3Aissue+milestone%3A3.5.0) for it. Not that a think doing it now is a bad idea, which is why I was testing it, and is why I found this crash.

**Comment by ivan-mogilko** [comment]
Looks like I had a false memory of opening this issue. I wrote the draft plan back on forums last summer, but never did the proper ticket. I only opened few complementary ones like #541 and #542.
Tbh I was first planning to try this on my own, but when you started working in this direction I assumed you are aiming similar thing. Then got distracted and lost the focus of this task.

I think we may announce 3.5.0 a beta version soon, after which we don't plan anything big in this release, but since it may take more time for people to test it and for us to fix bugs in existing features, we could as well try resolving last issues in rebuilding spritefile. If not, like I said, maybe I could find time to complete it as an additional update later...

This is why I wanted to ask you to open tickets on remaining problems and tasks connected to this (such as a need of storing checksum), but forgot to do this earlier. For example, there's one thing from the listed above which purpose I don't understand at all:
> project load process should be possible without reading any sprite data

PS. Regarding #541, that was one of the reasons I backported room code refactoring from 4.0 into 3.5.0. And there's another utility function that extracts a room component, so I believe it should be relatively easy to write similar function which reads room palette out if necessary.

**Comment by morganwillcock** [comment]
> project load process should be possible without reading any sprite data (otherwise you have to switch in a placeholder just to get the point where the editor is functional

i.e. if you don't have a spritecache file to start with, you cannot open the project to initiate its recreation.

Sprites are being enumerated when the project opens, regardless of there being no editor panels active at this time.

**Comment by ivan-mogilko** [comment]
>  if you don't have a spritecache file to start with, you cannot open the project to initiate its recreation.

Oh, right, I remember this annoyed me at some point. I think this is also the reason we ship default spritefile with AGS.

**Comment by ericoporto** [comment]
What was the conclusion from above discussion? This issue sounds important.

**Comment by ivan-mogilko** [comment]
Solution depends on whether a) fix is done in current version, where _native_ spritecache is used at design time, or b) we wait until next major version where _native_ spritecache is no more in the editor, and sprite file only created as an output.

In case a), the fix would be to continiously resave sprite file, freeing sprites from cache, if it becomes too big. This should be done both in case of regular sprite imports, and in case of reimport operation.

In case b, it depends on how sprite management works (but I guess it could be similar thing written in C#).

**Comment by ericoporto** [comment]
For b, this would be the editor following the image paths itself and rendering the images directly. This is not as trivial when tiled sprite import is taking into account. I am not sure but maybe the Editor uses the native proxy for GUI and room Editor too - in other places, this seems easier to change  (like the sprite and view panels...). If removing native proxy is a desired thing an issue should be created to plan this.

Question: could the editor easily use multiple spritecaches (say they top at 1GB...) And have they noted somewhere (like a map, sprite id to spritecache... Maybe it could be concatenated on save, and have the map deleted on success) ? And what is the problem of saving the big spritecache in "chunks"?

**Comment by ivan-mogilko** [comment]
> Question: could the editor easily use multiple spritecaches (say they top at 1GB...)

Maybe, at the cost that will make addressing sprites more difficult, and native code (reused by the editor) will have to be changed accordingly. But what is the point?

> And what is the problem of saving the big spritecache in "chunks"? 

It's not about saving it in chunks. It's about saving it *periodically*, when it gathered too many sprites in memory, and releasing accumulated resources.

**Comment by ericoporto** [comment]
> Maybe, at the cost that will make addressing sprites more difficult, and native code (reused by the editor) will have to be changed accordingly. But what is the point?

Better not do this, I don't have interest in heavily change the native code.

> It's not about saving it in chunks. It's about saving it periodically, when it gathered too many sprites in memory, and releasing accumulated resources.

Yes that sounds good.

**Comment by ericoporto** [comment]
@morganwillcock , do you still have your branch around? Could you link it here?

**Comment by morganwillcock** [comment]
For the spritecache rebuild it is added by the last two commits here:
https://github.com/morganwillcock/ags/tree/rebuild-spritecache

...but I think when I was testing this problem I was just importing a large amount of sprites at once and not using this branch. The memory limit is much lower than 4GB, at a guess it might be because there is a 2GB limit because of the default linking option for a 32 bit build:
https://docs.microsoft.com/en-us/cpp/build/reference/largeaddressaware-handle-large-addresses

**Comment by ericoporto** [comment]
Uhm, about ram memory, instead of using that flag, it seems easier to just build the Editor for 64-bits. The only thing in my head would be that using 32 bit plugins in Windows would require having the header pre-extracted or using 64 bit version of said plugin - but this would imply having 64 build of the engine like is done in the ports.

**Comment by ivan-mogilko** [comment]
I was running into "Out of mem" error simply by importing very large 32-bit bitmaps (like 5000x5000) and not saving. After about 5-6 sprites editor crashes.
But of course this is slightly different problem, because if you forcedly save spritefile, that means you are stealing user's ability to undo things. Which means for this particular case one would perhaps have to design a system of saving into temp files......
If we stick strictly to fixing this for reimport, then things become easier.

Moving to 64-bit may help to *some users* to avoid this problem temporarily, but that's just putting more burden on machine. And, plus, editor is still restricted to what memory user's system have, 64-bit program won't increase it's memory.

**Comment by ericoporto** [comment]
It should probably be easier to fix this while being MORE restricted in RAM (using 32 bits) since it will help the outofmemory problem happen sooner.

But what I mentioned was that we can know how much RAM a bitmap will take in memory, before loading the bitmap (height * width * bit depth), so what I thought was giving the option to save beforehand. And yes, temp files could work too, but in my head there's no way to concatenate files in-place, so it would need to create a copy of the full cache before the import.

**Comment by ivan-mogilko** [comment]
> , but in my head there's no way to concatenate files in-place, so it would need to create a copy of the full cache before the import.

Could you elaborate on that? What do you mean by "concatenate files in place", and why would we need this?

**Comment by ericoporto** [comment]
> Could you elaborate on that?

Sure. Let's say instead of importing to existing spritecache I impor to a new file, so the generated bitmap data goes to this new file. Now I save, I copy the data from this new file, at the end of the previous existing file, and also add the new index information. This should be fine but for a moment it would occupy twice the size of the added sprites in the disk - but now that I think about this, this should be fine. So yes, I think I was confused. In my mind that copy was expensive in size, but I was thinking about copying the entirety of the spritecache instead of just the added bits.

**Comment by morganwillcock** [comment]
I'm not sure that it is feasible to try to calculate safe memory usage because there is no guarantee on how fast the garbage collector will release it. For the same reason I don't think it is possible to warn or know when running out of memory.

> That would work as a quick fix.

I'm not aware of any users actually hitting the limitation, although admittedly crash reports would be useful here to confirm this is the case. Just increasing the memory limit will help and might be reasonably easy to do, maybe this is enough of a quick fix for ags3, and the actual solution to the problem is the separation of the build tools.

**Comment by ivan-mogilko** [comment]
> I'm not sure that it is feasible to try to calculate safe memory usage because there is no guarantee on how fast the garbage collector will release it.

I thought we were talking about sprites allocated by native means?

**Comment by ivan-mogilko** [comment]
Just to clarify, I'm not against expanding program memory capability of course, as it may be useful on its own (so long as it does wont suddenly cut off part of user base who dont fit new system requirements).

**Comment by morganwillcock** [comment]
> I thought we were talking about sprites allocated by native means? Unless managed applications wrap native allocations too somehow, which i am not aware of.

I don't think it is 100% native (at the very least the native part knows nothing about cropping tiles from a spritesheet), but according to memory usage reported by Windows the point where it crashed seemed to vary a lot. If the limit is actually 2GB I don't think I ever got near that, for me I think it was always between 1.1GB and 1.5GB.

**Comment by ericoporto** [comment]
If you are at 1.6 and try to import 0.4001 it should crash at 1.6 (failing for out of memory). Recovering from an out of memory is hard - can't even show a proper error since I need memory to allocate the message I wish to display. It sounds for me easier to predict in some way. If we can do with holding 1GB at the memory, than we should not try to allocate more than this and instead write on disk, be warning the user first and then writing on the same file, or writing to a new file and figure out how to concatenate later.

**Comment by ivan-mogilko** [comment]
> It can be done writing to a vector of bytes and counting the index, having a top limit - idnk, MAX_BYTES = 1024*1024*1024.

Sprite cache does about this thing.

**Comment by ivan-mogilko** [comment]
> It can be done writing to a vector of bytes and counting the index, having a top limit 

Sprite cache is supposed to calculate how much memory sprites in it have. It may require some amendments though, as I just noticed that it does not adds to the counter in case of sprites added in certain way. There might be a way to distinguish sprites that are not serialized using flags (it already has some), and keep track of their size separately.

**Comment by ericoporto** [comment]
That would work. As long as we set a smaller memory budget then one that would lead to Out of Memory errors or similar, it's possible to work with that in mind having the disk as the place to stash what has been in memory.

**Comment by ericoporto** [comment]
> There might be a way to distinguish sprites that are not serialized using flags

Hey, do you know if flag below does this? I was having a hard time following the code from managed to native (VS wasn't setting the breakpoints for some reason), but I wondered if this flag did this because of the comment in the line.

https://github.com/adventuregamestudio/ags/blob/2312e405c442f3bba1f69711fea392603cb6e910/Common/ac/spritecache.cpp#L163

**Comment by ivan-mogilko** [comment]
No, it's SPRCACHEFLAG_ISASSET, here's list of flags with comments:
https://github.com/adventuregamestudio/ags/blob/2312e405c442f3bba1f69711fea392603cb6e910/Common/ac/spritecache.h#L49

**Comment by ericoporto** [comment]
Ah, I understand now, whenever a **new** sprite is added to the cache (one NOT from asset file), it gets the `SPRCACHEFLAG_LOCKED` flag. This includes the recently imported ones in the editor.

But once the spritecache is saved, after saving, it reloads whatever was saved to disk, and at load time all Flags are reset - and at this point it loses the locked property - it becomes from the asset file!

**Comment by ericoporto** [comment]
I found three ways (there are probably more, these are my findings) to fix this:

**A) save in chunks**

- cons: current save requires knowing if spritecache is compressed and such save is written on `AGSEditor.cs` and also spread to other files (but not SpriteTools)
- need some rework so saving the spritecache can be moved to `Utils/SpriteTools.cs`, which is where sprite import and replace happens.
- doesn't require changing `spritecache.cpp`

---

**B) spritecache Import Sprite aware**
- assume spritecache works like a small sprite DB instead of cache
- Instead of reusing SetSprite, create a new function for ImportSprite.
- ImportSprite writes directly to disk, but to a temporary file. Imported sprites are not locked, need new fields to store new information.
- It should be possible to retrieve a sprite from a temporary file so it can be read.
- At save, read the sprites from temporary file and add them to the main asset file. Remove the temporary file afterwards.
- Not easy to modify, may leave files on temp dir occasionally, but doesn't change how spritecache functions used by the engine works.

---

**C) allow unload of locked sprites to a "swapfile"**

- locked sprites can be written to temporary location as a last resort if all other sprites that could be released have been.
- they keep a reference to the disk location, and asking for them allows to read the bmp blob on disk Instead.
- saving deletes the temporary sprites on disk
- add some other way to delete the sprites from temporary location
- a new flag is needed to indicate this case
- modifies engine functions so may have implications when running games.

**Comment by ivan-mogilko** [comment]
@ericoporto could you please clarify what does "save in chunks" mean, because I cannot understand exact mechanism from the description. I.e. "save in chunks" may be "save into several files", or "save in several attempts" etc, maybe something else?

I might also mention that we do not have to keep same class/mechanism shared between editor and engine of course. So if editor requires something different, we may as well split the class, make a subtype, or completely different class, etc. One of the issues with the spritecache class at the moment is that it combines both sprite file handling and sprite cache itself, which is not good.

Another thing to keep in mind, there's an idea to completely change how editor works with sprites, as in - let it read them right from the directory and not save to spritefile until the game is compiled. That involves writing completely separate sprite management for the editor. There's no ticket for this yet, but it should be and this might be a future goal for ags4. If we take that route, it may be not good to do alot of changes to the current system.

**Comment by ericoporto** [comment]
I added a small edit above in A) to clarify what I mean. I worry if import only happens at compile time it will make compiling slow.

**Comment by persn** [comment]
> there's this unmerged PR from Morgan that doesn't remove but lessen the spritecache dependency on the Editor

I will probably do something similar when I finish the open room format and move over to the sprite cache revamp. But I'll probably do it more thoroughly, I imagine a dedicated service class for doing sprite stuff similar to the `IRoomController` rather than a utility method which the PR uses. I want to see an issue that describes the topic though I dive into it and start imagining production code.

**Comment by rofl0r** [comment]
> I imagine a dedicated service class for doing sprite stuff similar to the IRoomController rather than a utility method which the PR uses.

you might want to read [this paper from Sony R&D](http://harmful.cat-v.org/software/OO_programming/_pdf/Pitfalls_of_Object_Oriented_Programming_GCAP_09.pdf) before you start with this, since it outlines the reasons for why the spritecache currently works the way it does.

**Comment by ivan-mogilko** [comment]
@rofl0r this issue is related to the editor and editing process, not engine and reading resources at runtime.

**Comment by ivan-mogilko** [comment]
Also, to be honest, I do not see how above paper is explaining current spritecache. I did not read it very thoroughly, but from a quick glance (and because I've already read similar topics before) my guess is it's speaking about benefits of continious data allocations. If it is so, the spritecache is not really continious, because each bitmap is allocated separately. Only some meta data is continious there (and data like sprite widths and heights afaik is duplicated elsewhere in the engine).

**Comment by rofl0r** [comment]
well, sprite index was clearly added as a speed optimization, and having all the width, heights and offsets packed into a single block allows to avoid cache misses when processing multiple sprites (of course that same cache isn't used when reading the actual, most often huge, imagedata).

**Comment by ivan-mogilko** [comment]
Yes, in regards to that it makes sense of course.

**Comment by ivan-mogilko** [comment]
Made a ticket for ags4: #1281

**Comment by ericoporto** [comment]
I have been thinking on something like this:
Original images + import rules -> individual bmp blobs -> make the spritecache file.

The individual bmp blobs would be akin to the sound cache files, they would be the import result but already in disk. 

This way if the imported images result in bigger than 2 GB (limit in 32 bit) it won't fail for lack of memory when writing the sprite cache file.

**Comment by ivan-mogilko** [comment]
@ericoporto are you speaking about what to do in ags4, or what to do in ags3?

How the editor will work with these "blobs" (read and write them)?

**Comment by ivan-mogilko** [comment]
BTW, I probably should mention, I was [splitting the sprite cache code](565eaaf90e63ad444798cd949b1e4c60dfb4a0f8), into SpriteCache and SpriteFile classes, where SpriteFile is working with file stream obviously, and SpriteCache, well, caching. It works more or less the same in a big picture, except as code is more separated it may be easier to change further later if necessary.

Anyway, one of the side effects of this work (also unplanned): when resaving sprite file with different compression mode the old code would actually load all the sprites into cache first. In the new code it still loads the sprites, but discards them immediately after, so this hopefully reduces memory use for this particular situation.

**Comment by ivan-mogilko** [comment]
@ericoporto I've been thinking more about your most recent proposal, the big issue is that current cache system streams from 1 file and is based on a list of offsets. If we save each new sprite in a separate file we would also have to rewrite the cache mechanism, making it work with a list of files instead. That will be somewhat close to what is proposed for ags4 (#1281), but made in native code. I'm concerned we will end up having two big works in parallel, one of which will eventually replace another anyway.

In the context of ags3, the dumbest solution is to introduce some size limit after which the Editor will tell user it cannot add more sprites until sprite file is saved. This is most straightforward, although perhaps not convenient to user (but better inconvenient than crashing with "out of memory" loosing your changes), and does not change overall work process.

---

On another hand, your last proposal gave me a hypothetical idea, which I'm thinking over now. It's about having a temp sprite file which is being constantly appended until user saves the project, after which the contents of a temp file are copied over real file. That would mean that there are no "unsaved" sprites.

Because current cache is working with offsets, the sprites order actually does not have to be sequential *so long as there's offsets table separately*. This means that, for example, if user replaces sprite in the middle, we do not have to shift all the file data, but instead may change the *offset* parameter of that sprite and append to the end instead.

But this is pretty vague at the moment, and involves several potential problems.

**Comment by ericoporto** [comment]
What I meant by bmp blobs is the actual data that we place in the spritecache, which is NOT a valid bmp, it has no headers for example.  But your suggestion might work too, but you will find some trouble similar to what I found in case of A in my previous comment here: https://github.com/adventuregamestudio/ags/issues/702#issuecomment-844091901

> It's about having a temp sprite file which is being constantly appended until user saves the project

To me this looks like the only feasible way of doing this without having too much complex code. There's still a lot of moving code around in the Editor, but yeah, it would work.

**Comment by ivan-mogilko** [comment]
## UPDATE + SUMMARY

### To reiterate on the problem:

1) The *newly imported* sprites accumulate in program memory and are only written to disk when user hits Save to save whole project. At some point the amount of memory these sprites take may exceed the available program space, and program crashes.
2) What contributes to this: *new sprites* and *reimported sprites* (including reloaded from source file).
3) When this may be a problem in practice: when working on high-res games, importing large number of sprites in batch, or reimporting large batch from source files.

### An update on situation

After resolving full recreation of a sprite file (#1404) as a separate process which is _much less_ prone to running out of mem, this particular issue is perhaps *somewhat* less important, although still may be met randomly when working with high-resolution games. 

Like prior, I believe that the true solution to this is changing the sprite storage in the project (#1281) - that would also resolve multiple issues at once. I also do not think it's a good idea to spend time inventing anything complicated, such as a different temporary storage format, as an intermediate solution in 3.* branch. If that seems really necessary to fix in 3.*, there were few *less complex* solutions mentioned in this thread:

1) Detect when editor is near the memory limit, warn the user and suggest to save the spritefile, thus dumping all the recently imported sprites from memory to disk, before proceeding further. This may be done even in the midst of a batch import / reimport.
2) Detect when editor is near the memory limit, automatically dump the sprites into the temporary spritefile, and switch cache reading to the same spritefile. When user orders to save the project, move that temp spritefile replacing the actual one. Editor will likely still have to notify the user about the saving, as this process may take a while for larger projects.
3) Similarily, have the temp spritefile, but write into it continuously as the new sprites are added, and let the cache read from the same file too. This will require to change the rule that the sprites in a file go in a direct order of IDs, and rely strictly on the offset table. The offset table currently is saved in a separate "sprite index" file, so that file might have to be updated and resized dynamically, but it's smaller, so will be a much faster operation. As the user saves the project, the temp file must be "finalized" by regrouping all sprites in a proper order and removing disposed sprites and replaced duplicates.
